cmd/validate.go

```diff
@@ -0,0 +1,272 @@
+}
+
+func validateMain(cmd *cobra.Command, args []string) exitCode {
+	// Validate that 'plugin' is a valid plugin and load it
+	plugins, serverOpts, err := serverOptsFor(cmd)
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	parallel, err := cmd.Flags().GetInt("parallel")
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	all, err := cmd.Flags().GetBool("all")
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	plug := args[0]
+	root, ok := plugins[plug]
+	if !ok {
+		// See if it's a script we can run as an external plugin instead
+		spec := plugin.ExternalPluginSpec{Script: plug}
+		root, err = spec.Load()
+		if err != nil {
+			pluginNames := make([]string, 0, len(plugins))
+			for name := range plugins {
+				pluginNames = append(pluginNames, name)
+			}
+			msg := "Unable to load %v as a configured plugin or script: %v\nValid plugins are: %v\n"
+			cmdutil.ErrPrintf(msg, plug, err, strings.Join(pluginNames, ", "))
+			return exitCode{1}
+		}
+	}
+
+	// Configure logging
+	log.SetFormatter(&log.TextFormatter{DisableTimestamp: true})
+	logFH, err := serverOpts.SetupLogging()
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+	if logFH != nil {
+		defer logFH.Close()
+	}
+
+	registry := plugin.NewRegistry()
+	if err := registry.RegisterPlugin(root, serverOpts.PluginConfig[plug]); err != nil {
+		cmdutil.ErrPrintf("Error loading plugin: %v\n", err)
+		return exitCode{1}
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	plugin.InitCache()
+	ctx := context.Background()
+
+	pw := progress.NewWriter()
+	pw.SetUpdateFrequency(50 * time.Millisecond)
+	pw.Style().Colors = progress.StyleColorsExample
+	go pw.Render()
+
+	// Use list to walk the hierarchy breadth-first. If a schema is present, visit each unique type
+	// at each level once, and stop when we've fully explored the schema. If there are errors, print
+	// all the detail we have about them. Use an unbuffered channel to register errors as they come in.
+	erred := 0
+	errs := make(chan error)
+	go func() {
+		for err := range errs {
+			erred++
+			cmdutil.ErrPrintf("%v\n", err)
+		}
+	}()
+
+	// We use a worker pool to limit work-in-progress, and a wait group to know when all queued work
+	// is complete. Put the plugin on the worker pool. Listing the registry should never error.
+	wp := newPool(parallel)
+	entries, _ := plugin.CachedList(ctx, registry)
+	for _, e := range entries {
+		wp.Submit(func() { processEntry(ctx, pw, wp, e, all, errs) })
+	}
+
+	// Wait for the workgroup's we've queued to finish, then stop the worker pool.
+	wp.Finish()
+	// Leave time for progress to finish rendering.
+	time.Sleep(100 * time.Millisecond)
+	pw.Stop()
+	if erred > 0 {
+		cmdutil.ErrPrintf("Found %v errors.\n", erred)
+		return exitCode{1}
+	}
+
+	cmdutil.Println("Looks good!")
+	return exitCode{0}
+}
+
+type pool struct {
+	wp *workerpool.WorkerPool
+	wg *sync.WaitGroup
```

Comment:
Comment at line 117 explains it.
> We use a worker pool to limit work-in-progress, and a wait group to know when all queued work is complete.

Reply:
Oh ok. I was asking because the wording in [StopWait's docs](https://godoc.org/github.com/gammazero/workerpool#WorkerPool.StopWait) suggests that  `StopWait` already does that, but I'm not at all familiar with the library.

Reply:
I need to allow new tasks to be submitted. I don't know how many tasks there will be, so don't know when to stop without having something that keeps track of all the potential tasks in the system.

Reply:
Ah ok. So the waitGroup's there b/c the queue's dynamic, while `StopWait()` assumes a static queue, right? If so, a comment along those lines would be appreciated